WordPress/src/main/java/org/wordpress/android/ui/posts/reactnative/ReactNativeRequestHandler.kt

```diff
@@ -0,0 +1,69 @@
+package org.wordpress.android.ui.posts.reactnative
+
+import androidx.core.util.Consumer
+import kotlinx.coroutines.CoroutineDispatcher
+import kotlinx.coroutines.GlobalScope
+import kotlinx.coroutines.launch
+import org.wordpress.android.fluxc.model.SiteModel
+import org.wordpress.android.fluxc.store.ReactNativeFetchResponse
+import org.wordpress.android.fluxc.store.ReactNativeFetchResponse.Error
+import org.wordpress.android.fluxc.store.ReactNativeFetchResponse.Success
+import org.wordpress.android.fluxc.store.ReactNativeStore
+import org.wordpress.android.modules.BG_THREAD
+import javax.inject.Inject
+import javax.inject.Named
+
+class ReactNativeRequestHandler @Inject constructor(
+    private val reactNativeStore: ReactNativeStore,
+    @Named(BG_THREAD) private val bgDispatcher: CoroutineDispatcher
+) {
+    fun performGetRequest(
+        pathWithParams: String,
+        mSite: SiteModel,
+        onSuccess: Consumer<String>,
+        onError: Consumer<String>
+    ) {
+        GlobalScope.launch(bgDispatcher) {
+            if (mSite.isUsingWpComRestApi) {
+                performGetRequestForWPComSite(pathWithParams, mSite.siteId, onSuccess::accept, onError::accept)
+            } else {
+                performGetRequestForSelfHostedSite(pathWithParams, mSite.url, onSuccess::accept, onError::accept)
+            }
+        }
+    }
+
+    private suspend fun performGetRequestForWPComSite(
+        pathWithParams: String,
+        wpComSiteId: Long,
+        onSuccess: (String) -> Unit,
+        onError: (String) -> Unit
+    ) {
+        parseUrlAndParamsForWPCom(pathWithParams, wpComSiteId)?.let { (url, params) ->
+            val response = reactNativeStore.performWPComRequest(url, params)
+            handleResponse(response, onSuccess, onError)
+        }
+    }
+
+    private suspend fun performGetRequestForSelfHostedSite(
+        pathWithParams: String,
+        siteUrl: String,
+        onSuccess: (String) -> Unit,
+        onError: (String) -> Unit
+    ) {
+        parseUrlAndParamsForWPOrg(pathWithParams, siteUrl)?.let { (url, params) ->
+            val response = reactNativeStore.performWPAPIRequest(url, params)
+            handleResponse(response, onSuccess, onError)
+        }
+    }
+
+    private fun handleResponse(
+        response: ReactNativeFetchResponse,
+        onSuccess: (String) -> Unit,
+        onError: (String) -> Unit
+    ) {
```

Comment:
I think this should have `withContext(mainDispatcher)` so you make sure the it's called on the main thread (it's injected with `@Named(UI_THREAD) private val mainDispatcher: CoroutineDispatcher`)

Reply:
Because this class is specifically for use only with React Native I'm not sure I agree. Anything sent to React Native is sent asynchronously across the javascript bridge, so keeping it on a background thread on the native side makes sense to me. If we ever were to try to do some UI updates on the native side based on the result of this call (I can't imagine why we would, but never say never 😄 ), I assume the issue would immediately present itself pretty clearly (immediate crash with UI-update-from-background-thread type of exception), and we would be better equipped to determine the best way to handle that issue at that time.

With that said, there's really not any processing done between this point and the bridge, so I don't necessarily think it would do any harm to move it to the main thread here. I just don't want to make that change if this is only going to be used with React Native and there's no benefit for that (React Native) use case. But what do you think?

Reply:
That makes sense. I just assumed that the callback is somehow affecting the UI and that's why I suggested the shift (basically the first `launch` jumps to the background thread so I assumed we want to go back to the original thread here).